File: bot/src/command/AdminsCommand.ts
import Command from "./Command";
import {VkMessage} from "../service/VkMessagesService";

export default class AdminsCommand extends Command {
    canYouHandleThisCommand(command: string, message: VkMessage): boolean {
        return command == "admins";
    }

    getCommandShortUsage(): string {
        return "/sweet admins (...)";
    }

    requiresPrivileges(peerId: number): boolean {
        return true;
    }

    async handle(command: string, rawArguments: string, message: VkMessage): Promise<void> {
        const { vkMessagesService } = this.context;

        const subCommand = rawArguments.split(" ")[0];
        const restArgs = rawArguments.substring(subCommand.length + 1);
        if (!subCommand) {
            await this.handleList(message);
        } else if (subCommand == 'add') {
            await this.handleAdd(restArgs, message);
        } else if (subCommand == 'remove') {
            await this.handleRemove(restArgs, message);
        } else {
            await vkMessagesService.send(message.peerId, this.getUsage());
        }
    }

    private async handleList(message: VkMessage): Promise<void> {
        const { vkMessagesService, chatAdminsOrmService } = this.context;
        const chatMembers = await vkMessagesService.getChatMembers(message.peerId);
        const sweetAdmins = await chatAdminsOrmService.getChatAdmins(message.peerId);
        let response = `Над Сладеньким имеют власть:\n`;
        if (sweetAdmins.length == 0) {
            response += `Никто :(`;
        } else {
            sweetAdmins.forEach(adminId => {
                const member = chatMembers.find(member => member.memberId == adminId);
                const displayName = member?.displayName || ('id' + adminId);
                response += `- ${displayName}\n`;
            });
        }
        await vkMessagesService.send(message.peerId, response);
    }

    private async handleAdd(args: string, message: VkMessage): Promise<void> {
        const { vkMessagesService, chatAdminsOrmService } = this.context;
        if (args.length == 0) {
            await vkMessagesService.send(message.peerId, this.getUsage());
            return;
        }
        const memberId = this.extractMemberId(args);
        if (memberId == null) {
            await vkMessagesService.send(message.peerId, `Не могу найти такого пользователя`);
            return;
        }
        const sweetAdmins = await chatAdminsOrmService.getChatAdmins(message.peerId);
        if (sweetAdmins.includes(memberId)) {
            await vkMessagesService.send(message.peerId, `Уже админ`);
            return;
        }
        await chatAdminsOrmService.addAdmin(message.peerId, memberId);
        await vkMessagesService.send(message.peerId, `Добавил админа`);
    }

    private async handleRemove(args: string, message: VkMessage): Promise<void> {
        const { vkMessagesService, chatAdminsOrmService } = this.context;
        if (args.length == 0) {
            await vkMessagesService.send(message.peerId, this.getUsage());
            return;
        }
        const memberId = this.extractMemberId(args);
        if (memberId == null) {
            await vkMessagesService.send(message.peerId, `Не могу найти такого пользователя`);
            return;
        }
        await chatAdminsOrmService.removeAdmin(message.peerId, memberId);
        await vkMessagesService.send(message.peerId, `Удалил админа`);
    }

    private extractMemberId(text: string): number | null {
        const regex = /\[?(id|club)(\d+)\|?/;
        const match = text.match(regex);
        if (match && !isNaN(+match[2])) {
            return +match[2];
        } else {
            return null;
        }
    }

    private getUsage(): string {
        let result = '';
        result += 'Команды:\n';
        result += '/sweet admins\n';
        result += '/sweet admins help\n';
        result += '/sweet admins add (тег)\n';
        result += '/sweet admins remove (тег)\n';
        return result;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/command/AdminsCommand.kt
package space.davids_digital.sweetie.command

import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.integration.vk.VkMessageService
import space.davids_digital.sweetie.integration.vk.VkUtils
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.orm.service.ChatAdminOrmService

@Component
class AdminsCommand(
    private val vkMessageService: VkMessageService,
    private val chatAdminOrmService: ChatAdminOrmService
): Command {
    companion object {
        private val log = LoggerFactory.getLogger(AdminsCommand::class.java)
    }

    override fun getNames(): Array<String> {
        return arrayOf("admins", "admin")
    }

    override fun getUsage(): String {
        return "admins (...)"
    }

    override fun requiresChatAdmin(): Boolean {
        return true
    }

    override fun requiresAppCeo(): Boolean {
        return false
    }

    @Transactional
    override suspend fun handle(commandName: String, rawArguments: String, message: VkMessageModel) {
        val subCommand = rawArguments.split(" ").firstOrNull() ?: ""
        val rest = rawArguments.substringAfter(subCommand)
        when (subCommand) {
            "" -> handleList(message)
            "add" -> handleAdd(message, rest)
            "remove" -> handleRemove(message, rest)
            else -> handleHelp(message)
        }
    }

    private fun handleList(message: VkMessageModel) {
        val adminIds = chatAdminOrmService.getAdminUserIds(message.peerId)
        val chatMembers = vkMessageService.getChatMembers(message.peerId)
        val result = StringBuilder()
        result.append("Над Сладеньким имеют власть:\n")
        adminIds.forEach { adminId ->
            val member = chatMembers.profiles.find { it.id == adminId }
            val group = chatMembers.groups.find { it.id == adminId }
            if (member != null) {
                result.append("- ${member.firstName} ${member.lastName}\n")
            } else if (group != null) {
                result.append("- ${group.name}\n")
            } else {
                result.append("- id$adminId\n")
            }
        }
        if (adminIds.isEmpty()) {
            result.append("Никто :(")
        }
        vkMessageService.send(message.peerId, result.toString())
    }

    private fun handleAdd(message: VkMessageModel, rest: String) {
        if (rest.isBlank()) {
            return handleHelp(message)
        }
        val memberId = VkUtils.extractMemberId(rest)
        if (memberId == null) {
            vkMessageService.send(message.peerId, "Не могу найти участника")
            return
        }
        val sweetAdmins = chatAdminOrmService.getAdminUserIds(message.peerId)
        if (sweetAdmins.contains(memberId)) {
            vkMessageService.send(message.peerId, "Уже админ")
            return
        }
        log.info("Adding admin $memberId to chat ${message.peerId}")
        chatAdminOrmService.addAdmin(message.peerId, memberId)
        vkMessageService.send(message.peerId, "Ок")
    }

    private fun handleRemove(message: VkMessageModel, rest: String) {
        if (rest.isBlank()) {
            return handleHelp(message)
        }
        val memberId = VkUtils.extractMemberId(rest)
        if (memberId == null) {
            vkMessageService.send(message.peerId, "Не могу найти участника")
            return
        }
        log.info("Removing admin $memberId from chat ${message.peerId}")
        chatAdminOrmService.removeAdmin(message.peerId, memberId)
        vkMessageService.send(message.peerId, "Ок")
    }

    private fun handleHelp(message: VkMessageModel) {
        vkMessageService.send(message.peerId, """
            Команды:
            /sweet admins
            /sweet admins help
            /sweet admins add (тег)
            /sweet admins remove (тег)
        """.trimIndent())
    }
}